### ISSUE TITLE
Thread-unsafe logic in krb5_init_context

### ISSUE THREAD

**Issue by riastradh** [heimdal/heimdal#1128]
**Describe the bug**
- krb5_init_context calls krb5_set_config_files at https://github.com/heimdal/heimdal/blob/9cbafd7320b8023ea8c6ea89643803bd021adda5/lib/krb5/context.c#L469
- which calls init_context_from_config_file at https://github.com/heimdal/heimdal/blob/9cbafd7320b8023ea8c6ea89643803bd021adda5/lib/krb5/context.c#L702
- which, if [libdefaults] dns_proxy is set, calls roken_gethostby_setup at https://github.com/heimdal/heimdal/blob/9cbafd7320b8023ea8c6ea89643803bd021adda5/lib/krb5/context.c#L190-L194
- which calls setup_int at https://github.com/heimdal/heimdal/blob/9cbafd7320b8023ea8c6ea89643803bd021adda5/lib/roken/roken_gethostby.c#L136
- which reads, modifies, and writes the dns_addr and dns_req objects at https://github.com/heimdal/heimdal/blob/9cbafd7320b8023ea8c6ea89643803bd021adda5/lib/roken/roken_gethostby.c#L66-L83
- which are defined with static (global) storage at https://github.com/heimdal/heimdal/blob/9cbafd7320b8023ea8c6ea89643803bd021adda5/lib/roken/roken_gethostby.c#L41-L42

Thus, if [libdefaults] dns_proxy is set, krb5_init_context must be serialized by calling threads, and two krb5_contexts can't be safely used at the same time if they have different config files with different settings of [libdefaults] dns_proxy.

It seems to me this should just use state inside the krb5_context, since the only user of the state, in roken_gethostby via roken_gethostbyname, is gethostname_fallback at https://github.com/heimdal/heimdal/blob/9cbafd7320b8023ea8c6ea89643803bd021adda5/lib/krb5/get_addrs.c#L58 which has access to the krb5_context.

(That said, I'm not sure it's worthwhile to have gethostname_fallback on modern operating systems, where I suspect the fallback will only be used in the event of running up against virtual address space ulimits causing malloc to fail, or on a system that has no network interfaces anyway.)

**To Reproduce**
code inspection of master as of 9cbafd7320b8023ea8c6ea89643803bd021adda5

**Expected behavior**
No races to access global shared state in the address space.

**Comment by nicowilliams** [comment]
We should audit for more of these globals.  These should be easy to fix by marking them thread-local.

**Comment by nicowilliams** [comment]
> I'm not sure making this thread-local is enough.  What happens if an application creates two contexts with different config files (or different realms with different DNS proxy settings, perhaps), and then interleaves their use in a single thread?

Indeed.  In this case I think we need a DNS resolver wrapper context.  Though this DNS HTTP proxy thing seems like a bit of a hack -- why not use HTTPS or TLS?  Well, I don't want to get into the business of maintaining a DNS resolver library, so perhaps we should just remove this code.

**Comment by nicowilliams** [comment]
I'm strongly inclined to remove it.

**Comment by nicowilliams** [comment]
Perhaps it could be useful for testing?  But we could mock DNS in other ways.
